### PR TITLE
refactor: query context

### PR DESCRIPTION
**Pull request by shuiyisong**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This pr mainly refactors `QueryContext`
1. change builder to use `derive_builder`
2. make `catalog` and `schema` within query context immutable
  - stateless clients like grpc and http should be able to change it on demand before sending request
  - stateful clients like mysql and pg should use `Session` to hold `catalog` and `schema`, and construct immutable query context on receiving next query
3. mark `arc()` as test only. Query context should be constructed with catalog and schema explicitly to avoid potential bug

Discussion
1. now that we actually use session to hold mutable `catalog` and `schema`, should we remove `Statement::Use` since its basically unreachable?
  - on mysql it calls `on_init` rather than treat `use` statement as a normal query
  - we require schema parameter on connecting with pg protocol. it would be strange to allow `use` statement within pg session; or we can process it before it goes into `SqlHandler` like mysql does

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by sunng87 — APPROVED** [review]

**Comment by sunng87** [comment]
> we require schema parameter on connecting with pg protocol. it would be strange to allow use statement within pg session; or we can process it before it goes into SqlHandler like mysql does

Sounds good to me. I think it's OK to have early termination for `USE` in postgresql handler

File: src/session/src/context.rs [hunk 1]
```diff
@@ -135,63 +98,29 @@ impl QueryContext {
...
+            time_zone: self.time_zone.unwrap_or(ArcSwap::new(Arc::new(None))),
+            sql_dialect: self.sql_dialect.unwrap_or(Box::new(GreptimeDbDialect {})),
+            trace_id: self.trace_id.unwrap_or(common_telemetry::gen_trace_id()),
+        }
     }
 
-    pub fn trace_id(mut self, trace_id: u64) -> Self {
-        self.trace_id = Some(trace_id);
-        self
+    pub fn build_to_arc(self) -> QueryContextRef {
```
Comment (MichaelScofield) [inline src/session/src/context.rs:116]: Can we just make the `build` method returned `QueryContextRef`?

File: src/frontend/src/statement.rs [hunk 2]
```diff
@@ -151,19 +150,13 @@ impl StatementExecutor {
...
-        ensure!(
-            self.catalog_manager
-                .schema_exist(catalog, &db)
-                .await
-                .context(CatalogSnafu)?,
-            SchemaNotFoundSnafu { schema_info: &db }
-        );
-
-        query_ctx.set_current_schema(&db);
+    async fn handle_use(&self, _db: String, _query_ctx: QueryContextRef) -> Result<Output> {
```
Comment (fengjiachun) [inline src/frontend/src/statement.rs:153]: Can this method be deleted?

**Comment by shuiyisong** [comment]
> > we require schema parameter on connecting with pg protocol. it would be strange to allow use statement within pg session; or we can process it before it goes into SqlHandler like mysql does
> 
> Sounds good to me. I think it's OK to have early termination for `USE` in postgresql handler

I've removed `handle_use` along with `Statement::Use`. Now if we try to execute `use` in pg client, it would responde
```
public=> show databases;
 Schemas
---------
 hello
 public
(2 rows)

public=> use hello;
ERROR:  Failed to execute query: use hello;, source: Failed to parse SQL, source: SQL statement is not supported: use hello;, keyword: use
```

File: src/session/src/context.rs [hunk 3]
```diff
@@ -53,66 +49,33 @@ impl Display for QueryContext {
...
+        QueryContextBuilder::default()
+            .current_catalog(catalog.to_string())
+            .current_schema(schema.to_string())
             .build()
     }
 
     #[inline]
     pub fn current_schema(&self) -> String {
-        self.current_schema.load().as_ref().clone()
+        self.current_schema.clone()
```
[the remark this answers is not in the document]
Comment (shuiyisong) [reply src/session/src/context.rs:66]: Upon first attempt, it feels nontrivial to make the change. `QueryContextRef` is an `Arc` struct we passed to functions using `clone()` and would be dropped at will. If `current_catalog()`(same as `current_schema()`) returns a reference, the returned value's lifetime seems to be bound to that temporary cloned `QueryContextRef`, making it vary short-lived and hard to use.

**Comment by shuiyisong** [comment]
@MichaelScofield @fengjiachun PTAL

**Review by MichaelScofield — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**